instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ConnectionSpanTest.java

```diff
@@ -0,0 +1,232 @@
+  }
+
+  int sendRequest(DefaultFullHttpRequest request, URI uri)
+      throws ExecutionException, InterruptedException, TimeoutException {
+    Channel channel = bootstrap.connect(uri.getHost(), uri.getPort()).sync().channel();
+    CompletableFuture<Integer> result = new CompletableFuture<>();
+    channel.pipeline().addLast(new ClientHandler(result));
+    channel.writeAndFlush(request).get();
+    return result.get(20, TimeUnit.SECONDS);
+  }
+
+  @Test
+  @DisplayName("test successful request")
+  void testSuccessfulRequest() throws Exception {
+    URI uri = URI.create("http://localhost:" + server.httpPort() + "/success");
+    DefaultFullHttpRequest request = buildRequest("GET", uri, new HashMap<>());
+    int responseCode = testing.runWithSpan("parent", () -> sendRequest(request, uri));
+
+    assertThat(responseCode).isEqualTo(200);
+    testing.waitAndAssertTraces(
+        trace ->
+            trace.hasSpansSatisfyingExactly(
+                span -> span.hasName("parent").hasKind(SpanKind.INTERNAL).hasNoParent(),
+                span ->
+                    span.hasName("RESOLVE")
+                        .hasKind(SpanKind.INTERNAL)
+                        .hasParent(trace.getSpan(0))
+                        .hasAttributesSatisfyingExactly(
+                            equalTo(ServerAttributes.SERVER_ADDRESS, uri.getHost()),
+                            equalTo(ServerAttributes.SERVER_PORT, uri.getPort())),
+                span ->
+                    span.hasName("CONNECT")
+                        .hasKind(SpanKind.INTERNAL)
+                        .hasParent(trace.getSpan(0))
+                        .hasAttributesSatisfyingExactly(
+                            equalTo(NetworkAttributes.NETWORK_TRANSPORT, "tcp"),
+                            equalTo(NetworkAttributes.NETWORK_TYPE, "ipv4"),
+                            equalTo(ServerAttributes.SERVER_ADDRESS, uri.getHost()),
+                            equalTo(ServerAttributes.SERVER_PORT, uri.getPort()),
+                            equalTo(NetworkAttributes.NETWORK_PEER_PORT, uri.getPort()),
+                            equalTo(NetworkAttributes.NETWORK_PEER_ADDRESS, "127.0.0.1")),
+                span -> span.hasName("GET").hasKind(SpanKind.CLIENT).hasParent(trace.getSpan(0)),
+                span ->
+                    span.hasName("test-http-server")
+                        .hasKind(SpanKind.SERVER)
+                        .hasParent(trace.getSpan(3))));
+  }
+
+  @Test
+  @DisplayName("test failing request")
+  void testFailingRequest() throws Exception {
+    URI uri = URI.create("http://localhost:" + PortUtils.UNUSABLE_PORT);
+    DefaultFullHttpRequest request = buildRequest("GET", uri, new HashMap<>());
+    Throwable thrownException = null;
+    try {
+      testing.runWithSpan("parent", () -> sendRequest(request, uri));
+    } catch (Throwable e) {
+      thrownException = e;
+    }
+
+    Throwable finalThrownException = thrownException;
```

Comment:
You could replace this block with:

```suggestion
    Throwable finalThrownException =
        catchThrowable(() -> testing.runWithSpan("parent", () -> sendRequest(request, uri)));
```

And then below in the assertions you can get rid of the `.hasEventsSatisfyingExactly` blocks and replace them with

```java
.hasException(finalThrownException),
```